src/renderers/dom/shared/utils/isCustomComponent.js

```diff
@@ -7,12 +7,38 @@
  * of patent rights can be found in the PATENTS file in the same directory.
  *
  * @providesModule isCustomComponent
+ * @flow
  */
 
 'use strict';
 
-function isCustomComponent(tagName, props) {
-  return tagName.indexOf('-') >= 0 || props.is != null;
+var DOMNamespaces = require('DOMNamespaces');
+var invariant = require('fbjs/lib/invariant');
+var HTML_NAMESPACE = DOMNamespaces.Namespaces.html;
+
+function isCustomComponent(
+  tagName: string,
+  props: Object,
+  domElement: Element | null,
```

Comment:
Seems like passing Element was a mistake that has lead to some inelegant code such as on the server where we don't have one. Don't pass around more than needed to complex functions.

Let's refactor this to pass what you need instead. Also if you need to check isCustomComponent in a bunch of places, just pass that as a Boolean instead of calling it multiple times. Nicer inverted control flow.

Reply:
I think we already pass it wherever possible. It's only called at entry points like diff and createInstance where we don't know that info.

Reply:
We shouldn't. It makes a lot of things harder such as server rendering and moving around object creation and hydration.

In this pr I don't understand what these functions do in the server renderer.

Reply:
They're needed to determine if we should follow the custom element path ("pass everything through without restrictions!") or the normal path. Do you mean you want to unify them?

Reply:
I believe this will be easier to refactor once Stack is removed. Part of the complexity is that `ReactDOMUnknownProperty` hooks are fired before a component is mounted. Stack doesn't assign a namespaceURI at that point. The current solution is a work around that, conveniently leans on the DOM element not being created yet in Stack to preserve the old custom element behavior.